packages/@ourworldindata/grapher/src/lineLegend/LineLegendHelpers.ts

```diff
@@ -0,0 +1,286 @@
+        if (numPicked === maxCandidatesToPick) break
+    }
+
+    return context
+}
+
+/**
+ * Pick as many candidates as possible from a given subset.
+ *
+ * The order of candidates to consider for placement is determined by the
+ * `popCandidateFromSubset` function. The function should return the next candidate
+ * to consider. If the function returns `undefined`, the algorithm stops.
+ *
+ * If no custom function is provided, the algorithm picks candidates starting from
+ * the end (!) of the given list.
+ */
+export function pickAsManyAsPossibleWithRetry(
+    params: PickFromCandidateSubsetParams
+): LineLegendFilterAlgorithmContext {
+    return pickFromCandidateSubsetWithRetry(params)
+}
+
+/**
+ * Pick a fixed number of candidates from a give subset.
+ *
+ * The order of candidates to consider for placement is determined by the
+ * `popCandidateFromSubset` function. The function should return the next candidate
+ * to consider. If the function returns `undefined`, the algorithm stops.
+ *
+ * If no custom function is provided, the algorithm picks candidates starting from
+ * the end (!) of the given list.
+ */
+export function pickCandidateWithRetry(
+    params: PickFromCandidateSubsetParams
+): LineLegendFilterAlgorithmContext {
+    return pickFromCandidateSubsetWithRetry({
+        ...params,
+        maxCandidatesToPick: 1,
+    })
+}
+
+export function pickCandidateWithMaxDistanceToReferenceCandidate(params: {
+    context: LineLegendFilterAlgorithmContext
+    candidateSubset: PlacedSeries[]
+    referenceCandidate: PlacedSeries
+}): LineLegendFilterAlgorithmContext {
+    const { context, candidateSubset, referenceCandidate } = params
+    const distanceMap = new Map(
+        candidateSubset.map((c) => [c.seriesName, dist(c, referenceCandidate)])
+    )
+    const pop = (candidates: PlacedSeries[]) =>
+        maxBy(candidates, (c) => distanceMap.get(c.seriesName))
+
+    return pickCandidateWithRetry({
+        context,
+        candidateSubset,
+        popCandidateFromSubset: pop,
+    })
+}
+
+export function pickOutermostCanidate(params: {
+    context: LineLegendFilterAlgorithmContext
+    candidateSubset: PlacedSeries[]
+}): LineLegendFilterAlgorithmContext {
+    const { context, candidateSubset } = params
+
+    if (candidateSubset.length === 0) return context
+
+    if (candidateSubset.length === 1) {
+        return pickCandidate(context, candidateSubset[1])
+    }
+
+    const minCandidate = minBy(candidateSubset, (c) => c.midY)!
+    const maxCanidate = maxBy(candidateSubset, (c) => c.midY)!
+    const distanceMap = new Map(
+        candidateSubset.map((c) => [
+            c.seriesName,
+            Math.min(dist(minCandidate, c), dist(maxCanidate, c)),
+        ])
+    )
+
+    const pop = (candidates: PlacedSeries[]) =>
+        minBy(candidates, (c) => distanceMap.get(c.seriesName))
+
+    return pickCandidateWithRetry({
+        context,
+        candidateSubset,
+        popCandidateFromSubset: pop,
+    })
```

Comment:
Okay, so I'm having a bit of a hard time understanding this method.

IIUC, it always picks candidates with `c.midY === minCandidate.midY || c.midY === maxCandidate.midY`.
Which can either be `minCandidate, maxCandidate` themselves, or any other candidates that share the same `midY`.
But it's doing this in quite a convoluted way, and the only slight advantage I can see is that it's supposedly gonna pick the first element in the list that fits this criterion.
Is there something I'm missing?

Reply:
ah, no, you're right, this is more complicated than it needs to be. The function used to pick two candidates at the same time and then I rewrote it but forgot to drop all that complexity...